vm/src/vm/vm_memory/memory_segments.rs

```diff
@@ -213,6 +213,25 @@ impl MemorySegmentManager {
         Ok(memory_holes)
     }
 
+    /// Returns a list of addresses of memory cells that constitute the public memory.
+    /// segment_offsets is the result of self.relocate_segments()
+    pub fn get_public_memory_addresses(&self, segment_offsets: &[usize]) -> Vec<(usize, &usize)> {
+        let len = self.num_segments().min(self.public_memory_offsets.len());
+        let mut addresses = Vec::with_capacity(len);
+
+        for (offsets, segment_start) in self
+            .public_memory_offsets
+            .values()
+            .zip(segment_offsets.iter())
```

Comment:
You can used `cloned()` on this iterator and just save yourself a lot of dealing with references:
```suggestion
            .zip(segment_offsets.iter().cloned())
```

Reply:
I'm not sure why is this a problem, though I could clone `page_id` to solve the weird `(usize, &usize)` return type that you mentioned in another review

Reply:
References leak into your structures and make life harder in the long term (you end up having to add lifetime parameters in several places). We use them because copies are typically expensive, but when we're talking about `usize` it's actually cheaper to just copy the value.

Reply:
That's great to know! thanks, but would it be applicable in this case? I'm using references only for the sum operation (`segment_start + offset`), whose result will be moved into the tuple that I'm pushing to the vec. Then at the end I'm returning a vector of tuples of values, not references. Am I missing something here?

Reply:
No problem! thanks for the good info nevertheless